armory/postprocessing/plot_patch_aware_carla_metric.py

```diff
@@ -0,0 +1,232 @@
+"""
+Utility functions for visualizing the output of the metric "object_detection_AP_per_class_by_giou_from_patch."
+This metric captures how adversarial AP varies as objects get farther from the patch.
+
+There are two functions which produce slightly different plots:
+plot_mAP_by_giou_with_patch, and plot_single_giou_threshold.
+
+plot_mAP_by_giou_with_patch() shows how the mAP changes over a range of GIoU values.  It
+can display this in three "flavors": cumulative by max GIoU, cumulative by min GIoU,
+or as a histogram.  The first flavor represents all objects outside each given range of GIoU
+(i.e. further from the patch, or more negative GIoU).  The second flavor is all objects within
+each range (i.e. closer to the patch, or more positive GIoU).  The histogram version reports AP
+for disjoint intervals of GIoU value.
+
+plot_single_giou_threshold() only considers one user-specified GIoU threshold, and shows
+the adversarial and benign AP both above and below that threshold.  For example, using a threshold
+of 0.0 would correspond to all objects touching and not touching the patch.
+
+Intended for stand-alone usage:
+>>> from armory.postprocessing.plot_patch_aware_carla_metric import plot_mAP_by_giou_with_patch
+>>> plot_mAP_by_giou_with_patch("path/to/results.json", flavors=["cumulative_by_min_giou"])
+
+"""
+import json
+
+from matplotlib import pyplot as plt
+import numpy as np
+
+
+ben_color = "tab:blue"
+adv_color = "tab:red"
+
+flavor_titles = {
+    "cumulative_by_max_giou": "Cumulative by Upper Bound\n(Exclusive)",
+    "cumulative_by_min_giou": "Cumulative by Lower Bound\n(Inclusive)",
+    "histogram_left": "Histogram\n(Left Oriented)",
+}
+
+fontsize = 8
+
+
+def _init_plots(json_filepath, include_classes, n_flavors=1):
+    # Read json from json_filepath and initialize fig and ax objects for plotting.
+
+    # Returns giou dicts from json, the fig and axes, as well as number of classes.
+
+    with open(json_filepath) as f:
+        blob = json.load(f)
+        results = blob["results"]
+
+    adv_giou = results["adversarial_object_detection_AP_per_class_by_giou_from_patch"][
+        0
+    ]
+    ben_giou = results["benign_object_detection_AP_per_class_by_giou_from_patch"][0]
+    adv_ap = results["adversarial_carla_od_AP_per_class"][0]["mean"]
+    ben_ap = results["benign_carla_od_AP_per_class"][0]["mean"]
+
+    if include_classes:
+        n_class = len(ben_giou["cumulative_by_max_giou"]["0.0"]["class"])
+    else:
+        n_class = 0
+
+    fig, axes = plt.subplots(n_class + 1, n_flavors, sharex=True, sharey=True)
+    # Ensure axes are in an array for consistent reference
+    if type(axes) != np.array or len(axes.shape) == 1:
```

Comment:
Addressing https://github.com/twosixlabs/armory/actions/runs/5788881976/job/15688781545?pr=1946
```suggestion
    if isinstance(axes, np.array) or len(axes.shape) == 1:
```